File: bukkit/src/test/java/ru/endlesscode/mimic/bukkit/BukkitSystemRegistryTest.java
/*
 * This file is part of BukkitMimic.
 * Copyright (C) 2018 Osip Fatkullin
 * Copyright (C) 2018 EndlessCode Group and contributors
 *
 * BukkitMimic is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BukkitMimic is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with BukkitMimic.  If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.mimic.bukkit;

import org.bukkit.plugin.ServicePriority;
import org.bukkit.plugin.ServicesManager;
import org.junit.Before;
import org.junit.Test;
import ru.endlesscode.mimic.api.system.ClassSystem;
import ru.endlesscode.mimic.api.system.LevelSystem;
import ru.endlesscode.mimic.api.system.PlayerSystem;
import ru.endlesscode.mimic.api.system.SystemFactory;
import ru.endlesscode.mimic.api.system.registry.SubsystemPriority;
import ru.endlesscode.mimic.bukkit.system.PermissionsClassSystem;
import ru.endlesscode.mimic.bukkit.system.TestSystem;
import ru.endlesscode.mimic.bukkit.system.VanillaLevelSystem;

import static org.junit.Assert.*;

public class BukkitSystemRegistryTest extends BukkitTestBase {
    private BukkitSystemRegistry systemRegistry;

    @Before
    public void setUp() {
        super.setUp();

        ServicesManager servicesManager = this.server.getServicesManager();
        servicesManager.unregisterAll(plugin);

        this.systemRegistry = new BukkitSystemRegistry(plugin, servicesManager);
    }

    @Test
    public void testRegisterRightSubsystemMustBeSuccessful() throws Exception {
        this.systemRegistry.registerSubsystem(VanillaLevelSystem.class, VanillaLevelSystem.FACTORY);
        this.systemRegistry.registerSubsystem(PermissionsClassSystem.class, PermissionsClassSystem.FACTORY);

        SystemFactory<LevelSystem> levelSystemFactory = this.systemRegistry.getSystemFactory(LevelSystem.class);
        SystemFactory<ClassSystem> classSystemFactory = this.systemRegistry.getSystemFactory(ClassSystem.class);

        assertNotNull("System must be initialized", levelSystemFactory);
        assertNotNull("System must be initialized", classSystemFactory);
    }

    @Test
    public void testGetNotRegisteredSystemMustReturnNull() throws Exception {
        checkSystemNotExists(TestSystem.class);
    }

    @Test
    public void testGetServicePriorityFromSystem() {
        assertEquals(ServicePriority.Lowest, BukkitSystemRegistry.servicePriorityFromSystem(SubsystemPriority.LOWEST));
        assertEquals(ServicePriority.Low, BukkitSystemRegistry.servicePriorityFromSystem(SubsystemPriority.LOW));
        assertEquals(ServicePriority.Normal, BukkitSystemRegistry.servicePriorityFromSystem(SubsystemPriority.NORMAL));
        assertEquals(ServicePriority.High, BukkitSystemRegistry.servicePriorityFromSystem(SubsystemPriority.HIGH));
        assertEquals(ServicePriority.Highest, BukkitSystemRegistry.servicePriorityFromSystem(SubsystemPriority.HIGHEST));
    }

    @Test
    public void testUnregisterAllSubsystemsMustBeSuccessful() throws Exception {
        this.systemRegistry.registerSubsystem(VanillaLevelSystem.class, VanillaLevelSystem.FACTORY);
        this.systemRegistry.registerSubsystem(PermissionsClassSystem.class, PermissionsClassSystem.FACTORY);

        this.systemRegistry.unregisterAllSubsystems();

        checkSystemNotExists(LevelSystem.class);
        checkSystemNotExists(ClassSystem.class);
    }

    @Test
    public void testUnregisterExistingFactoryMustBeSuccessful() throws Exception {
        this.systemRegistry.registerSubsystem(PermissionsClassSystem.class);
        this.systemRegistry.unregisterFactory(PermissionsClassSystem.FACTORY);

        checkSystemNotExists(ClassSystem.class);
    }

    @Test
    public void testUnregisterNotExistingFactoryMustBeSuccessful() {
        this.systemRegistry.unregisterFactory(PermissionsClassSystem.FACTORY);

        checkSystemNotExists(ClassSystem.class);
    }

    private <SystemT extends PlayerSystem> void checkSystemNotExists(Class<SystemT> systemClass) {
        assertNull(this.systemRegistry.getSystemFactory(systemClass));
    }
}

File: bukkit/src/test/kotlin/ru/endlesscode/mimic/bukkit/BukkitSystemRegistryTest.kt
/*
 * This file is part of BukkitMimic.
 * Copyright (C) 2018 Osip Fatkullin
 * Copyright (C) 2018 EndlessCode Group and contributors
 *
 * BukkitMimic is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BukkitMimic is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with BukkitMimic.  If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.mimic.bukkit

import ru.endlesscode.mimic.api.system.ClassSystem
import ru.endlesscode.mimic.api.system.LevelSystem
import ru.endlesscode.mimic.api.system.PlayerSystem
import ru.endlesscode.mimic.api.system.registry.getSystemFactory
import ru.endlesscode.mimic.api.system.registry.registerSubsystem
import ru.endlesscode.mimic.bukkit.system.vanilla.PermissionsClassSystem
import ru.endlesscode.mimic.bukkit.system.vanilla.VanillaLevelSystem
import kotlin.test.*

class BukkitSystemRegistryTest : BukkitTestBase() {

    // SUT
    private lateinit var systemRegistry: BukkitSystemRegistry

    @BeforeTest
    override fun setUp() {
        super.setUp()

        systemRegistry = BukkitSystemRegistry(plugin, server.servicesManager)
    }

    @Test
    fun `when register correct subsystem - should register it`() {
        // When
        val registered = systemRegistry.registerSubsystem(PermissionsClassSystem.FACTORY)

        // Then
        assertTrue(registered)
        assertNotNull(systemRegistry.getSystemFactory<ClassSystem>())
    }

    @Test
    fun `when no system registered - should return null for any system`() {
        // Then
        checkSystemNotExists<LevelSystem>()
    }

    @Test
    fun `when unregister all subsystems - should return null for all previously registered subsystems`() {
        // Given
        systemRegistry.registerSubsystem(VanillaLevelSystem.FACTORY)
        systemRegistry.registerSubsystem(PermissionsClassSystem.FACTORY)

        // When
        systemRegistry.unregisterAllSubsystems()

        // Then
        checkSystemNotExists<LevelSystem>()
        checkSystemNotExists<ClassSystem>()
    }

    @Test
    fun `when unregister factory - should unregister subsystem`() {
        // Given
        systemRegistry.registerSubsystem<PermissionsClassSystem>()

        // When
        systemRegistry.unregisterFactory(PermissionsClassSystem.FACTORY)

        // Then
        checkSystemNotExists<ClassSystem>()
    }

    @Test
    fun `when unregister not registered subsystem - should not throw exception`() {
        // When
        systemRegistry.unregisterFactory(PermissionsClassSystem.FACTORY)

        // Then
        checkSystemNotExists<ClassSystem>()
    }

    @AfterTest
    fun tearDown() {
        server.servicesManager.unregisterAll(plugin)
    }

    private inline fun <reified SystemT : PlayerSystem> checkSystemNotExists() {
        assertNull(systemRegistry.getSystemFactory<SystemT>())
    }
}
